src/d3-org-chart.js

```diff
@@ -678,9 +715,134 @@ export class OrgChart {
             attrs.firstDraw = false;
         }
 
+        attrs.svg = svg;
+
+        // // add drag and drop
+        d3.attrs = attrs;
+
         return this;
     }
 
+    dragAttachHandler() {
+        const attrs = this.getChartState();
+        attrs.svg.selectAll(".node").call(
+            d3.drag()
+                .on("start", this.dragstarted)
+                .on("drag", this.dragged)
+                .on("end", this.dragended)
+        );
+    }
+    dragstarted(d) {
+        console.log(d)
+        d.sourceEvent.stopPropagation();
+        d3.select(this).classed("dragging", true);
+        d3.sourceNode = d;
+    }
+    dragged(d, event) {
+        const x = (d.x) - (event.width / 2);
+
+        // const _x = (event.x - d.x);
+        // const _y = (event.y - d.y);
+        // const moveThreshold = 30;
+        // const isMoved = (_x > -(moveThreshold) && _x < (moveThreshold)) || (_y > -(moveThreshold) && _y < (moveThreshold));
+
+        // if(!isMoved) return;
+
+        d3.select(this).raise().attr("transform", `translate(${x},${d.y})`);
+        // set default style
+        d3.selectAll("rect").attr("fill", "#fff").attr("stroke", "null").attr(
+            "stroke-width",
+            "1px"
+        );
+        d3.targetNode = null;
+        // check nodes overlapping
+        const cP = {
+            x0: d.x,
+            y0: d.y,
+            x1: d.x + event.width,
+            y1: d.y + event.height,
+        };
+
+        d3.selectAll("g.node:not(.dragging)").filter((d, i) => {
+            const cPInner = {
+                x0: d.x,
+                y0: d.y,
+                x1: d.x + d.width,
+                y1: d.y + d.height,
+            };
+            if (
+                (cP.x1 > cPInner.x0 && cP.x0 < cPInner.x1) &&
+                (cP.y1 > cPInner.y0 && cP.y0 < cPInner.y1)
```

Comment:
I ran into issues here where a node would have collisions with multiple nodes. 
Ideally we want a collision with one node. 

So in my code I changed this condition to account for the width of the node. 

Basically if the sourceNode is halfway over the  targetNode I then assign it as the target node. 

```ts
cP.x1 - event.width / 2 > cPInner.x0 &&
cP.x0 + event.width / 2 < cPInner.x1 &&
cP.y1 - event.height / 2 > cPInner.y0 &&
cP.y0 + event.height / 2 < cPInner.y1
```

Reply:
You could also store halfWidth and halfHeight on cP so that you don't have to keep calculating inside the condition

Reply:
```ts
const cP = {
    x0: d.x,
    y0: d.y,
    x1: d.x + event.width,
    y1: d.y + event.height,
    midX: d.x + event.width / 2
    midY: d.y + event.height / 2
};



cP.midX > cPInner.x0 &&
cP.midX < cPInner.x1 &&
cP.midY > cPInner.y0 &&
cP.midY < cPInner.y1
```